volatility3/framework/symbols/windows/extensions/__init__.py

```diff
@@ -775,39 +776,131 @@ def get_peb(self) -> interfaces.objects.ObjectInterface:
-
         try:
-            peb = self.get_peb()
-            yield from peb.Ldr.InLoadOrderModuleList.to_list(
-                f"{self.get_symbol_table_name()}{constants.BANG}_LDR_DATA_TABLE_ENTRY",
-                "InLoadOrderLinks",
-            )
+            pebs = [
+                self.get_peb(),
+                self.get_peb32(),
+            ]
+            for peb in pebs:
+                if peb:
+                    sym_table = self.get_symbol_table_name()
+                    if peb.Ldr.vol.type_name.endswith("unsigned long"):
+                        ldr_data = self._context.symbol_space.get_type(
+                            self._32bit_table_name + constants.BANG + "_PEB_LDR_DATA"
+                        )
+                        peb.Ldr = peb.Ldr.cast("pointer", subtype=ldr_data)
+                        sym_table = self._32bit_table_name
+                    yield from peb.Ldr.InLoadOrderModuleList.to_list(
+                        f"{sym_table}{constants.BANG}" + "_LDR_DATA_TABLE_ENTRY",
+                        "InLoadOrderLinks",
+                    )
         except exceptions.InvalidAddressException:
             return None
 
     def init_order_modules(self) -> Iterable[interfaces.objects.ObjectInterface]:
         """Generator for DLLs in the order that they were initialized"""
 
         try:
-            peb = self.get_peb()
-            yield from peb.Ldr.InInitializationOrderModuleList.to_list(
-                f"{self.get_symbol_table_name()}{constants.BANG}_LDR_DATA_TABLE_ENTRY",
-                "InInitializationOrderLinks",
-            )
+            pebs = [
+                self.get_peb(),
+                self.get_peb32(),
+            ]
+            for peb in pebs:
+                if peb:
+                    sym_table = self.get_symbol_table_name()
+                    if peb.Ldr.vol.type_name.endswith("unsigned long"):
+                        ldr_data = self._context.symbol_space.get_type(
+                            self._32bit_table_name + constants.BANG + "_PEB_LDR_DATA"
+                        )
+                        peb.Ldr = peb.Ldr.cast("pointer", subtype=ldr_data)
+                        sym_table = self._32bit_table_name
+                    yield from peb.Ldr.InInitializationOrderModuleList.to_list(
+                        f"{sym_table}{constants.BANG}" + "_LDR_DATA_TABLE_ENTRY",
+                        "InInitializationOrderLinks",
+                    )
         except exceptions.InvalidAddressException:
             return None
 
     def mem_order_modules(self) -> Iterable[interfaces.objects.ObjectInterface]:
         """Generator for DLLs in the order that they appear in memory"""
-
         try:
-            peb = self.get_peb()
-            yield from peb.Ldr.InMemoryOrderModuleList.to_list(
-                f"{self.get_symbol_table_name()}{constants.BANG}_LDR_DATA_TABLE_ENTRY",
-                "InMemoryOrderLinks",
-            )
+            pebs = [
```

Comment:
This feels like duplicating quite a big more code than the previous version.  It might be worth pulling out the pointer check on these into a separate little private subfunction?